planner/scan.go

```diff
@@ -60,17 +63,84 @@ func (n *scanNode) Kind() string {
 
 func (n *scanNode) Init() error {
 	// init the fetcher
-	if err := n.fetcher.Init(&n.desc, n.fields, n.reverse, n.showDeleted); err != nil {
+	if err := n.fetcher.Init(&n.desc, n.fields, n.filter, n.slct.DocumentMapping, n.reverse, n.showDeleted); err != nil {
 		return err
 	}
 	return n.initScan()
 }
 
 func (n *scanNode) initCollection(desc client.CollectionDescription) error {
 	n.desc = desc
+	n.initFields(n.slct.Fields)
+	if n.slct.OrderBy != nil {
+		for _, cond := range n.slct.OrderBy.Conditions {
+			for _, index := range cond.FieldIndexes {
+				name, _ := n.docMapper.documentMapping.TryToFindNameFromIndex(index)
+				n.tryAddField(name)
+			}
+		}
+	}
 	return nil
 }
 
+func (n *scanNode) initFields(fields []mapper.Requestable) {
+	for _, r := range fields {
+		// add all the possible base level fields the fetcher is responsible
+		// for, including those that are needed by higher level aggregates
+		// or grouping alls, which them selves might have further dependants
+		switch requestable := r.(type) {
+		// field is simple as its just a base level field
+		case *mapper.Field:
+			n.tryAddField(requestable.GetName())
+		// select might have its own select fields and filters fields
+		// @todo: Might need to check that these are sub selects
+		// are _groups
+		case *mapper.Select:
```

Comment:
question: This case looks like it should be impossible within the scanNode - are you really sure it ever gets run?  And if so when?

Reply:
Two instances right now. For regular type joins, it may need the coresponding <type>_id field. For `_groups` it needs to get the fields within the group selectionSet. Both of these are defined as `mapper.Select`s within a scan context.

Reply:
I did however remove the `filter` loop, as that was moved to the mapper after this change, so was able to trim some stuff.

Reply:
All the relevant fields on any child stuff, should be present at the parent level - all the needs you just described should be handled by the `case *mapper.Field:` case within this switch

Reply:
*Should be* as in that is how you believe it currently works, or.. *Should be* I should make the necessary changes to make this work?

Reply:
Both.  I believe it does this now.  And conceptually it should do that anyway - the whole point of mapper is to produce a data structure that is ready to use by planner so it doesnt have to bother doing any of this stuff and can instead focus on planner stuff.

Reply:
The mapper doesn't do this at the moment, as it was written when the fetcher would always return all fields, so even the most basic of selects would gurantee we get all dependent fields that any of the filters, orders, childs, groups, etc would ever need, so it didn't need to be overly accurate. All my changes to the mapper have been to add much of this functionality into the mapper.

In any case, the goal of the mapper is to provide a structured way to access information in the query/document using array indexes, hence *mapper* (mapping between maps and slices).

Don't think of this as mapper code in the planner, this collecting of fields would happen even if the mapper didn't exist, and we still used the old map based representation for things. The mapper structures all the necessary information (for the most part) so that it can be consumed in w.e way, but each respective node in the plan graph that consumes mapper structures has to grab its own specific information. The `scanNode` just has to grab the most, as there are various ways that dependent fields can sneak into what a query needs.

IMO this code isn't doing the job of the mapper, we still want the fields for a `_group` for example, within a child select type, because the `groupNode` needs it that way, or the `typeJoinNode`, etc. This is just pulling the now more structured and more explicit info from the `mapper.Select` into the scanNode and fetcher

Reply:
If a field isn't present in mapper.IndexesByName, it shouldn't be fetched/scanned, if it is, it should be.  That really is all the logic that should be here. I think you've even done most/all the work required on the mapper side.

Reply:
We had previously talked about updating the `IndexesByName` in the `core.DocumentMapping` object that gets passed around, but, for a bunch of different reasons I decided that wasn't the best approach. Most notably, the mapping types are passed around, so there aren't that many unique ones, where as there are several instances where the scanNode/fetcher are using the same collection, but for different reasons, with different fields, filters, etc. So it would be really awkward if the mapping is shared among all the `mapper.Select` types to try to convey this information via the `DocumentMapping`.

The `mapper.Select` are already unique, and contain the various specific filter/order information, so the goal was to just make use of that structure to convey all the information regarding dependent fields.

Reply:
Ah fair enough, I see where you are coming from now.  I think I would probably have still used a shared `IndexesByName` and over-fetched fields that may be required by one Requestable but not another, I think that the maintenance savings  would be worth the performance hit.

Happy as is then :)